### PR TITLE
Added support for (biblatex) `langid` to be an optional field in entry editor

### LINKED ISSUE #10868: Support for (biblatex) langid field in entry editor
JabRef 5.12

I might have overlooked it, but I do not find the rather central `langid` field in the entry editor. This field defines the principal language of the entry, which is important for hyphenation and whether (sentence) casing is applied. There is also `langidopts` for fine-tuning (with `polyglossia`)

The field is described in http://mirrors.ctan.org/macros/latex/contrib/biblatex/doc/biblatex.pdf at section 2.2.3

In my opinion, this should go to the "Optional" tab.

### PR DESCRIPTION
**Pull request by mag-sun**
## Overview:
Integrated 'langid' into all biblatex/bibtex supported entry types as an optional field. This is so formatting of entries can be adapted to the language of the entries. Closes #10868.

## Changes implemented:
- Added a 'Langid' enum class defining all valid language ids provided by the biblatex documentation ([https://au.mirrors.cicku.me/ctan/macros/latex/contrib/biblatex/doc/biblatex.pdf]). 
- Added a 'LanguageEditorViewModel' class to handle the gui and dropdown menu. 
- Introduced 'langid' as an important field in all relevant biblatex entry type definitions.

https://github.com/user-attachments/assets/acf87f19-94ab-4c54-8bd4-39717fa916dc

image

### Mandatory checks

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -57,6 +58,7 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 - When fulltext search is selected but indexing is deactivated, a dialog is now shown asking if the user wants to enable indexing now [#9491](https://github.com/JabRef/jabref/issues/9491)
 - We changed instances of 'Search Selected' to 'Search Pre-configured' in Web Search Preferences UI. [#11871](https://github.com/JabRef/jabref/pull/11871)
 - We added a new CSS style class `main-table` for the main table. [#11881](https://github.com/JabRef/jabref/pull/11881)
+- We changed BibTex and BibLatex entries to include `langid` as an optional field. [#10868](https://github.com/JabRef/jabref/issues/10868)
```
Comment (koppor) [inline CHANGELOG.md:61]: One entry in the CHANGELOG.md is OK. remove this. (No need to be very exact in the CHANGELOG).

File: gradle.properties [hunk 2]
```diff
@@ -11,3 +11,5 @@ org.gradle.jvmargs=-Xmx6096M
 
 # hint by https://docs.gradle.org/current/userguide/performance.html#enable_the_build_cache
 org.gradle.caching=true
+org.gradle.parallel=true
+org.gradle.daemon=true
```
Comment (koppor) [inline gradle.properties:15]: Are you sure abot this?

We should have parallel tests: https://github.com/JabRef/jabref/pull/9797

---

Please remove this - we do not know about any side effects.

File: src/main/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModel.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+//
+//    public void setValue(Langid langid) {
+//        selectedLangidProperty.set(langid);
+//    }
+//
+//    // Optionally add a property getter
+//    public ObjectProperty<Langid> selectedLangidProperty() {
+//        return selectedLangidProperty;
+//    }
+
```
Comment (koppor) [inline src/main/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModel.java:38]: Remove commented code

File: src/main/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModel.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import org.jabref.logic.integrity.FieldCheckers;
+import org.jabref.model.database.BibDatabaseMode;
+import org.jabref.model.entry.Langid;
+import org.jabref.model.entry.field.Field;
+import org.jabref.model.strings.StringUtil;
+
+public class LangidEditorViewModel extends OptionEditorViewModel<Langid> {
+    private BibDatabaseMode databaseMode;
+
+//    private final ObjectProperty<Langid> selectedLangidProperty = new SimpleObjectProperty<>();
```
Comment (koppor) [inline src/main/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModel.java:20]: Remove commented code

File: src/main/java/org/jabref/logic/msbib/MSBibEntry.java [hunk 5]
```diff
@@ -58,7 +58,8 @@ class MSBibEntry {
     public String albumTitle;
     public String broadcastTitle;
     public String year;
-    public String month;
+    public String
+            month;
```
Comment (koppor) [inline src/main/java/org/jabref/logic/msbib/MSBibEntry.java:62]: Please revert this.

File: src/main/java/org/jabref/model/entry/Langid.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
+package org.jabref.model.entry;
+
+import java.util.Optional;
+
+import org.jabref.logic.bibtex.FieldWriter;
+
+public enum Langid {
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/Langid.java:7]: Please reference the biblatex manual as source of these values. You can use `<a href ..`) for JAvaDoc or Markdown syntax (with three ///)

File: src/main/java/org/jabref/model/entry/field/FieldProperty.java [hunk 7]
```diff
@@ -18,7 +18,7 @@ public enum FieldProperty {
     IDENTIFIER,
 
     LANGUAGE,
-
+    LANGUAGEID,
     // Field content is text, but should be interpreted as markdown
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/field/FieldProperty.java:22]: Please keep empty lines as separator

File: src/main/java/org/jabref/model/entry/Langid.java (hunk 6 above)
Comment (koppor) [reply src/main/java/org/jabref/model/entry/Langid.java:7]: Can these values also be used for the `language` field?

File: src/main/java/org/jabref/model/entry/field/FieldProperty.java (hunk 7 above)
Comment (koppor) [reply src/main/java/org/jabref/model/entry/field/FieldProperty.java:22]: Maybe, this and `LANGUAGE` can just be removed. These properties are not used.

Alternataively, use it. And then use the langId field editor as you introduced for all fields having this property.

File: src/test/java/org/jabref/model/entry/LangidTest.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    @Test
+    public void testLangidNameAndValue() {
+        assertEquals("basque", Langid.BASQUE.getLangid());
+        assertEquals("Basque", Langid.BASQUE.getName());
+        assertEquals("american", Langid.AMERICAN.getLangid());
+        assertEquals("American", Langid.AMERICAN.getName());
+    }
+
```
Comment (koppor) [inline src/test/java/org/jabref/model/entry/LangidTest.java:89]: This is generatged, isn't it? Useless test IMHO. Remove

File: src/test/java/org/jabref/model/entry/LangidTest.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        return Stream.of(
+                arguments("Basque", Langid.BASQUE),
+                arguments("Bulgarian", Langid.BULGARIAN),
+                arguments("American", Langid.AMERICAN),  // Test for the langid value
+                arguments("British", Langid.BRITISH),
+                arguments("US English", Langid.USENGLISH),
+                arguments("Swiss German", Langid.SWISSGERMAN),
+                arguments("Norwegian (Bokmål)", Langid.NORSK)
+        );
+    }
```
Comment (koppor) [inline src/test/java/org/jabref/model/entry/LangidTest.java:106]: Remove

**Review by koppor — CHANGES_REQUESTED** [review]
Initial comments.

Please re-iterate on the biblatex smanual and check whether `language` and `origLanguage` also have the same values. If yes, your editor can be used if the field proprty FieldProperty.LANGUAGE langauge is present. If not, then this property can just be remoed.

In other words FieldPropety groups fields together and indicates the editor for it. Do not introduce a property if it used for one field only.

File: gradle.properties [hunk 10]
```diff
@@ -11,5 +11,4 @@ org.gradle.jvmargs=-Xmx6096M
 
 # hint by https://docs.gradle.org/current/userguide/performance.html#enable_the_build_cache
 org.gradle.caching=true
-org.gradle.parallel=true
```
Comment (Siedlerchr) [inline gradle.properties]: Do not remove this

File: src/test/java/org/jabref/model/entry/LangidTest.java [hunk 11]
```diff
@@ -78,30 +77,4 @@ private static Stream<Arguments> getLangidByStringTest() {
                 arguments(Optional.empty(), "unknown")
         );
     }
-
-    @Test
-    public void testLangidNameAndValue() {
```
Comment (Siedlerchr) [inline src/test/java/org/jabref/model/entry/LangidTest.java]: why do you remove these test?

**Review by Siedlerchr — CHANGES_REQUESTED** [review]

File: src/main/java/org/jabref/model/entry/types/BibtexEntryTypeDefinitions.java [hunk 12]
```diff
@@ -20,12 +20,12 @@ public class BibtexEntryTypeDefinitions {
      * An article from a journal or magazine.
      * <p>
      * Required fields: author, title, journal, year.
-     * Optional fields: volume, number, pages, month, note.
+     * Optional fields: volume, number, pages, month, note, langid.
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/types/BibtexEntryTypeDefinitions.java:23]: This is a biblatex field, not a BibTeX field. Please revert. --> `git checkout main -- src/main/java/org/jabref/model/entry/types/BibtexEntryTypeDefinitions.java`

File: src/main/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModel.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+import javafx.util.StringConverter;
+
+import org.jabref.gui.autocompleter.SuggestionProvider;
+import org.jabref.logic.integrity.FieldCheckers;
+import org.jabref.model.database.BibDatabaseMode;
+import org.jabref.model.entry.Langid;
+import org.jabref.model.entry.field.Field;
+import org.jabref.model.strings.StringUtil;
+
+public class LangidEditorViewModel extends OptionEditorViewModel<Langid> {
```
Comment (koppor) [inline src/main/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModel.java:17]: Call this `LanguageEditor...`, because it is related to the fieldproperty `Language` (and also for the field `language`).

File: src/main/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModel.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+            public String toString(Langid object) {
+                if (object == null) {
+                    return null;
+                } else {
+                    return object.getLangid();  // Langid used as both display and value
+                }
+            }
+
+            @Override
+            public Langid fromString(String string) {
```
Comment (koppor) [inline src/main/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModel.java:38]: Return `Optional<Langid>`  instead of `null`.

File: src/main/java/org/jabref/model/entry/field/StandardField.java [hunk 15]
```diff
@@ -78,7 +78,7 @@ public enum StandardField implements Field {
     KEY("key"),
     KEYWORDS("keywords"),
     LANGUAGE("language", FieldProperty.LANGUAGE),
-    LANGUAGEID("langid", FieldProperty.LANGUAGE),
+    LANGUAGEID("langid", FieldProperty.LANGUAGEID),
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/field/StandardField.java:81]: No, keep `FieldProperty.LANGUAGE`

File: src/test/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModelTest.java [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+    void getItemsShouldReturnAllLangidValues() {
+        Collection<Langid> items = langidEditorViewModel.getItems();
+        assertEquals(Langid.values().length, items.size());
+        assertTrue(items.contains(Langid.BASQUE)); // Check if it contains a specific Langid
+        assertTrue(items.contains(Langid.AMERICAN)); // Additional check for another Langid
+    }
+
+    @Test
+    void testStringConversion() {
+        // Test conversion from string to Langid
```
Comment (koppor) [inline src/test/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModelTest.java:60]: This is the method title - remove this comment.

File: src/test/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModelTest.java [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        String langidString = "bulgarian";
+        Langid langid = langidEditorViewModel.getStringConverter().fromString(langidString);
+        assertEquals(Langid.BULGARIAN, langid, "String should convert to the corresponding Langid");
+
+        // Test conversion from Langid to string
+        String convertedString = langidEditorViewModel.getStringConverter().toString(Langid.BULGARIAN);
+        assertEquals(langidString, convertedString, "Langid should convert back to its string representation");
```
Comment (koppor) [inline src/test/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModelTest.java:67]: Do a seperate test.

Remove the assert text, because this should be the test method name.

(Also do for the other tests)

File: CHANGELOG.md [hunk 18]
```diff
@@ -34,6 +34,7 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 - We added a different background color to the search bar to indicate when the search syntax is wrong. [#11658](https://github.com/JabRef/jabref/pull/11658)
 - We added a setting which always adds the literal "Cited on pages" text before each JStyle citation. [#11691](https://github.com/JabRef/jabref/pull/11732)
 - We added a new plain citation parser that uses LLMs. [#11825](https://github.com/JabRef/jabref/issues/11825)
+- We added a 'language id' (langid) field for biblatex entry types. [#10868](https://github.com/JabRef/jabref/issues/10868)
```
Comment (koppor) [inline CHANGELOG.md:37]:
```suggestion
- We added support for `langid` field for biblatex libraries. [#10868](https://github.com/JabRef/jabref/issues/10868)
```

**Review by koppor — CHANGES_REQUESTED** [review]
Tests are a bit too verbose and not confirming to the one-assertion-per-test rule. Please streamline.

Other comments inside.

File: src/main/java/org/jabref/model/entry/types/BibtexEntryTypeDefinitions.java (hunk 12 above)
Comment (mag-sun) [reply src/main/java/org/jabref/model/entry/types/BibtexEntryTypeDefinitions.java:23]: Yes my apologies, I was working in the BibTex class before noting the difference between the two

File: src/main/java/org/jabref/model/entry/field/FieldProperty.java (hunk 7 above)
Comment (mag-sun) [reply src/main/java/org/jabref/model/entry/field/FieldProperty.java:22]: I have removed `LANGUAGEID` and instead using `LANGUAGE` for langid field editor/view model. I am now assuming that the language IDs given in the biblatex doc are also all the supported languages.

File: src/main/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModel.java (hunk 14 above)
Comment (mag-sun) [reply src/main/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModel.java:38]: I was under the impression that it had to return `Langid` because the superclass `OptionEditorViewModel` requires `StringConverter<T>` to return <T> not Optional<T>?

File: gradle.properties [hunk 19]
```diff
@@ -11,3 +11,4 @@ org.gradle.jvmargs=-Xmx6096M
 
 # hint by https://docs.gradle.org/current/userguide/performance.html#enable_the_build_cache
 org.gradle.caching=true
+org.gradle.parallel=true
```
Comment (koppor) [inline gradle.properties:14]: Revert this change.

File: src/main/java/org/jabref/model/entry/Langid.java [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+
+import java.util.Optional;
+
+import org.jabref.logic.bibtex.FieldWriter;
+/**
+ * Language identifiers based on BibLaTeX manual specifications.
+ * See the BibLaTeX documentation for full details:
+ * <a href="http://mirrors.ctan.org/macros/latex/contrib/biblatex/doc/biblatex.pdfhangelo">BibLaTeX manual</a>
+ */
+
```
Comment (koppor) [inline src/main/java/org/jabref/model/entry/Langid.java:11]: JavaDoc needs to be attached to class --> remove empty line here (and add empty line above)

File: src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class LanguageEditorViewModelTest {
+
+    private LanguageEditorViewModel languageEditorViewModel;
+
+    @BeforeEach
+    void setUp() {
+        // Mock dependencies
```
Comment (koppor) [inline src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java:29]: Remove comment. The next lines state that by code.

File: src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+        BibDatabaseContext databaseContext = Mockito.mock(BibDatabaseContext.class);
+        FilePreferences filePreferences = Mockito.mock(FilePreferences.class);
+        JournalAbbreviationRepository abbreviationRepository = Mockito.mock(JournalAbbreviationRepository.class);
+
+        // Initialize FieldCheckers with mocked instances
```
Comment (koppor) [inline src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java:34]: Remove comment. The next lines state that by code.

File: src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+        FieldCheckers fieldCheckers = new FieldCheckers(databaseContext, filePreferences, abbreviationRepository, false);
+
+        // Mock the SuggestionProvider
```
Comment (koppor) [inline src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java:37]: Remove comment. The next lines state that by code.

File: src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+        SuggestionProvider<?> suggestionProvider = Mockito.mock(SuggestionProvider.class);
+
+        // Initialize the LangidEditorViewModel
```
Comment (koppor) [inline src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java:40]: Remove comment. The next lines state that by code.

Comments are useful if they add information. E.g. for the reasons of the code... - But not necessary here IMHO.

File: src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+        languageEditorViewModel = new LanguageEditorViewModel(
+                StandardField.LANGUAGEID,  // Use the correct field
```
Comment (koppor) [inline src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java:42]: Remove comment.

File: src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+                suggestionProvider,  // Mocked SuggestionProvider
```
Comment (koppor) [inline src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java:43]: Remove comment.

File: src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+                BibDatabaseMode.BIBLATEX,  // Use the correct BibDatabaseMode
```
Comment (koppor) [inline src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java:44]: Remove comment.

File: src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java [hunk 28]
```diff
@@ -0,0 +1,94 @@
...
+                fieldCheckers,  // FieldCheckers instance
```
Comment (koppor) [inline src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java:45]: Remove comment.

File: src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+                new UndoManager()  // UndoManager instance
```
Comment (koppor) [inline src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java:46]: Remove comment.

File: src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java [hunk 30]
```diff
@@ -0,0 +1,94 @@
...
+        );
+    }
+
+    @Test
+    void getItemsShouldReturnAllLangidValues() {
+        Collection<Langid> items = languageEditorViewModel.getItems();
+        assertEquals(Langid.values().length, items.size());
+        assertTrue(items.contains(Langid.BASQUE)); // Check if it contains a specific Langid
```
Comment (koppor) [inline src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java:54]: Remove comment.

File: src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java [hunk 31]
```diff
@@ -0,0 +1,94 @@
...
+        assertTrue(items.contains(Langid.AMERICAN)); // Additional check for another Langid
```
Comment (koppor) [inline src/test/java/org/jabref/gui/fieldeditors/optioneditors/LanguageEditorViewModelTest.java:55]: Try to rewrite multiple assert statements to one

assertEquals(List.of(...), items)

**Review by koppor — CHANGES_REQUESTED** [review]
Minor comments.

Tests have to be rewritten to fit JabRef style.

No text for failing assertions.

File: src/main/java/org/jabref/model/entry/Langid.java (hunk 20 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/model/entry/Langid.java:11]: Works with line in between

File: src/main/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModel.java (hunk 14 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/gui/fieldeditors/optioneditors/LangidEditorViewModel.java]: Yes, null is correct here.

**Review by Siedlerchr — DISMISSED** [review]

**Review by Siedlerchr — DISMISSED** [review]

**Review by Siedlerchr — APPROVED** [review]
